### PR TITLE
Next

### PR DESCRIPTION
**Pull request by kaste**
Closes #146 
Closes #147 
Closes #148 
Closes #149 

Because in the end #146 and parts of #147 are breaking, we need to put them behind a feature flag. I used `fast_timings` here. The default is for obvious reasons 'False'. I recommend to deprecate the old `DeferringTextTestRunner` because it has no advantage over the new one. 

I back-ported as many features as possible to the old one though. 

All users get:

- yield <condition> throws and sends its value back (part of #147)
- yield AWAIT_WORKER (part of #146)

Users opting 'fast_timings' get:

- Super-fast tests 😁 
- yield condition will not continue but err on timeout
- For functional test, allow running single-threaded. This will make coverage happy, and allows ticking the time using plain `yield`. I highly recommend using this if the app/plugin you're building allows it.

Notable fix

- `add/doCleanups` works

Why is the system now so fast?

We do not add implicit waiting time anymore. 

What to do if the test fail after opting 'fast_timings'?

Essentially, *add* explicit waiting time using one of the `yield` forms. E.g. for the failing GitSavvy tests (discussed in #146) it is enough to `yield AWAIT_WORKER` on `tearDown`. This is reasonable, and it doesn't make the tests as slow as before. The runtime of the GitSavvy test is still cut in half. (Note that AWAIT_WORKER is introduced by this PR, you usually used number guessing like `yield 300` et.al. before.)

### CODE REVIEW

**Comment by kaste** [comment]
We could add this to the `DeferrableTestCase` as a convenience fixture:

```
def run_single_threaded(self):
    original = sublime.set_timeout_async
    sublime.set_timeout_async = sublime.set_timeout
    self.addCleanup(partial(setattr, sublime, 'set_timeout_async', original))
```

**Comment by randy3k** [comment]
Thanks. This looks more promising.

**Comment by kaste** [comment]
We now have three options 'deferred', 'async', and 'fast_timings'. I think we can deprecated all of them. 'deferred=False' is slower than the current 'deferred=True' and has really limited functionality. (Since these are global switches, it is unlikely that a Sublime plugin *never* wants to wait for a Sublime doing something.)

'async=True' is not the way forward either. I can imagine that we run the framework on a separate thread, but again the feature set of the deferred TestCase is superior. But running the code under test on a 'third' thread which then delegates to Sublime's threads ... don't think that makes a lot of sense. As a tester you usually don't want more async, you want less.

File: unittesting/core/st3/fast_runner.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+DEFAULT_CONDITION_POLL_TIME = 17
+DEFAULT_CONDITION_TIMEOUT = 4000
+AWAIT_WORKER = 'AWAIT_WORKER'
+# Extract `set_timeout_async`, t.i. *avoid* late binding, in case a user
+# patches it
+run_on_worker = sublime.set_timeout_async
+
+
+class FastDeferringTextTestRunner(TextTestRunner):
```
Comment (randy3k) [inline unittesting/core/st3/fast_runner.py]: How does it different from `DeferringTextTestRunner` except `DEFAULT_CONDITION_POLL_TIME`?
It seems that they are almost identical.

**Comment by kaste** [comment]
You need to diff them, really

```diff
diff --git a/unittesting/core/st3/runner.py b/unittesting/core/st3/fast_runner.py
index dff1fbf..3785aa8 100644
--- a/unittesting/core/st3/runner.py
+++ b/unittesting/core/st3/fast_runner.py
@@ -10,13 +10,15 @@ def defer(delay, callback, *args, **kwargs):
     sublime.set_timeout(partial(callback, *args, **kwargs), delay)


+DEFAULT_CONDITION_POLL_TIME = 17
+DEFAULT_CONDITION_TIMEOUT = 4000
 AWAIT_WORKER = 'AWAIT_WORKER'
 # Extract `set_timeout_async`, t.i. *avoid* late binding, in case a user
 # patches it
 run_on_worker = sublime.set_timeout_async


-class DeferringTextTestRunner(TextTestRunner):
+class FastDeferringTextTestRunner(TextTestRunner):
     """This test runner runs tests in deferred slices."""

     def run(self, test):
@@ -48,7 +50,7 @@ class DeferringTextTestRunner(TextTestRunner):
                     startTestRun()
                 try:
                     deferred = test(result)
-                    defer(10, _continue_testing, deferred)
+                    _continue_testing(deferred)

                 except Exception as e:
                     _handle_error(e)
@@ -61,10 +63,10 @@ class DeferringTextTestRunner(TextTestRunner):
                     condition = deferred.send(send_value)

                 if callable(condition):
-                    defer(100, _wait_condition, deferred, condition)
+                    defer(0, _wait_condition, deferred, condition)
                 elif isinstance(condition, dict) and "condition" in condition and \
                         callable(condition["condition"]):
-                    period = condition.get("period", 100)
+                    period = condition.get("period", DEFAULT_CONDITION_POLL_TIME)
                     defer(period, _wait_condition, deferred, **condition)
                 elif isinstance(condition, int):
                     defer(condition, _continue_testing, deferred)
@@ -73,7 +75,7 @@ class DeferringTextTestRunner(TextTestRunner):
                         partial(defer, 0, _continue_testing, deferred)
                     )
                 else:
-                    defer(10, _continue_testing, deferred)
+                    defer(0, _continue_testing, deferred)

             except StopIteration:
                 _stop_testing()
@@ -82,21 +84,29 @@ class DeferringTextTestRunner(TextTestRunner):
             except Exception as e:
                 _handle_error(e)

-        def _wait_condition(deferred, condition, period=100, timeout=10000, start_time=None):
+        def _wait_condition(
+            deferred, condition,
+            period=DEFAULT_CONDITION_POLL_TIME,
+            timeout=DEFAULT_CONDITION_TIMEOUT,
+            start_time=None
+        ):
             if start_time is None:
                 start_time = time.time()

             try:
                 send_value = condition()
             except Exception as e:
-                defer(10, _continue_testing, deferred, throw_value=e)
+                _continue_testing(deferred, throw_value=e)
                 return

             if send_value:
-                defer(10, _continue_testing, deferred, send_value=send_value)
+                _continue_testing(deferred, send_value=send_value)
             elif (time.time() - start_time) * 1000 >= timeout:
-                self.stream.writeln("Condition timeout, continue anyway.")
-                defer(10, _continue_testing, deferred)
+                error = TimeoutError(
+                    'Condition not fulfilled within {:.2f} seconds'
+                    .format(timeout / 1000)
+                )
+                _continue_testing(deferred, throw_value=error)
             else:
                 defer(period, _wait_condition, deferred, condition, period, timeout, start_time)

@@ -154,4 +164,4 @@ class DeferringTextTestRunner(TextTestRunner):
             else:
                 self.stream.write("\n")

-        sublime.set_timeout(_start_testing, 10)
+        _start_testing()
```

**Comment by kaste** [comment]
We have quite a few places where we just continue sync. And we use '0' often which has different semantics bc it will also continue immediately after other _immediate_ tasks have been run. ('0' will probably not yield back to Sublime's task scheduler but just append this task to end of the current loop.)

**Comment by randy3k** [comment]
You are copying `FastDeferringTextTestRunner` instead of subclassing it from `DeferringTextTestRunner`. Is it because you want to deprecate `DeferringTextTestRunner` eventually?

File: unittesting/core/st3/fast_runner.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+                    defer(0, _wait_condition, deferred, condition)
+                elif isinstance(condition, dict) and "condition" in condition and \
+                        callable(condition["condition"]):
+                    period = condition.get("period", DEFAULT_CONDITION_POLL_TIME)
+                    defer(period, _wait_condition, deferred, **condition)
+                elif isinstance(condition, int):
+                    defer(condition, _continue_testing, deferred)
+                elif condition == AWAIT_WORKER:
+                    run_on_worker(
+                        partial(defer, 0, _continue_testing, deferred)
```
Comment (randy3k) [inline unittesting/core/st3/fast_runner.py]: Once a TestCase yields to AWAIT_WORKER, the whole test case will be executed in the worker thread.
Should we send the coroutine back to the main thread like this?

```
partial(defer, 0, lambda: sublime.set_timeout(_continue_testing), deferred)
```

Comment (kaste) [reply unittesting/core/st3/fast_runner.py]: I already do the double-hop here to get it back on the main thread.

This line evaluates to

```
timeout_async(lambda: timeout(lambda: continue))
^^^ await worker      ^^^ hop back    ^^^actually continue
```

but this quickly gets confusing, so we may need a test here

Comment (randy3k) [reply unittesting/core/st3/fast_runner.py]: Thanks. I missed the point that `defer` is really `set_timeout`.

**Comment by kaste** [comment]
Subclassing is the wrong approach here. We don't want to maintain two or three versions of it. We actually ship the old version because Package Control doesn't do the semantic versioning right. 

As I see it, usually you would tag it 2.0, and people would just stick with their 1.x until they're ready. But we don't have that in Sublime land. So we basically ship two versions. (That's the same thing as when IE10 (InternetExplorer) ships actually a version of IE8 and IE9.)

If we subclass, we can never guarantee that we do NOT break the old version.

**Comment by kaste** [comment]
To be honest, backporting some features to the old Runner was maybe a complete waste of my time.

**Comment by randy3k** [comment]
Ok. I argue we should only keep on version of the runners. It could be done by 

1. adding settings to specify `DEFAULT_CONDITION_POLL_TIME`, `DEFAULT_CONDITION_TIMEOUT`. Say `condition_poll_time` and `condition_timeout`.

2. adding a setting to control how long the coroutine is deferred, `yield_timeout` (default 10). If `yield_timeout` is 0, it implies the use of 
```
_continue_testing(deferred, throw_value=e)
```
instead of
```
defer(10, _continue_testing, deferred, send_value=send_value)
```

**Comment by kaste** [comment]
A plain `yield` does a `timeout(fn, 0)` this is not the same as calling `fn()` directly. It appends `fn` to the end of the current run and awaits the task queue. 

In the PR I sometimes change `defer(10, fn) -> defer(0, fn)`, but sometimes I do `defer(10, fn) -> fn()`. Note that `DeferrableTestSuite` often uses plain `yield` which I do not tackle right now. (I hope we can remove them completely in a next version.)

Also, I think these explosion of settings confuse users. There is actually no continuum of 0..10ms yield time. 0 has different semantics. For the poll time, you want it low but you probably don't want it to be faster than 60hz, so I took 17ms. 

I don't think users should experiment with different settings here. They should switch one setting and get what we have right now plus what we might have later in `DeferrableTestSuite`.  

Now, of course, I tried different things. But adding 10 if statements in one function, doesn't cut it, you need different implementation via subclassing or functional vodoo. I experimented a bit here, both subclassing and functional composition, in then end and to be honest it felt like Operette. 

The code here is already difficult to reason about (and for lesser experienced readers even more so), sharing reusable bits makes it more complicated bc you have more delegation and late binding. We usually oscillate between DRY and KISS, and here with that code at hand, it is not simple enough to make it DRY.

I hope you have a really good idea.

File: unittesting/core/st3/fast_runner.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            timeout=DEFAULT_CONDITION_TIMEOUT,
+            start_time=None
+        ):
+            if start_time is None:
+                start_time = time.time()
+
+            try:
+                send_value = condition()
+            except Exception as e:
+                _continue_testing(deferred, throw_value=e)
```
Comment (randy3k) [inline unittesting/core/st3/fast_runner.py]: Please help me to understand. Why we cannot do `defer(0, _continue_testing, deferred, throw_value=e)`?

File: unittesting/core/st3/fast_runner.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+                return
+
+            if send_value:
+                _continue_testing(deferred, send_value=send_value)
+            elif (time.time() - start_time) * 1000 >= timeout:
+                error = TimeoutError(
+                    'Condition not fulfilled within {:.2f} seconds'
+                    .format(timeout / 1000)
+                )
+                _continue_testing(deferred, throw_value=error)
```
Comment (randy3k) [inline unittesting/core/st3/fast_runner.py]: same here.

File: unittesting/core/st3/fast_runner.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+            if expectedFails:
+                infos.append("expected failures=%d" % expectedFails)
+            if unexpectedSuccesses:
+                infos.append("unexpected successes=%d" % unexpectedSuccesses)
+            if infos:
+                self.stream.writeln(" (%s)" % (", ".join(infos),))
+            else:
+                self.stream.write("\n")
+
+        _start_testing()
```
Comment (randy3k) [inline unittesting/core/st3/fast_runner.py]: Can we do `defer(0, _start_testing)`?

Comment (randy3k) [inline unittesting/core/st3/fast_runner.py]: If we could replace the function calls above by `defer(....)`, it is strong reason that we should use the setting `yield_timeout`.

File: unittesting/core/st3/fast_runner.py (hunk 3 above)
Comment (randy3k) [reply unittesting/core/st3/fast_runner.py]: Oh, was it because of the coverage issue?

Comment (kaste) [reply unittesting/core/st3/fast_runner.py]: Okay, you're trying to modify the behavior to make the generalization/abstraction easier. That's cheating.

Now, sure both runners *look* similar but the execution flow is completely different. For sync programs, we can look at conditional branching (e.g. ifs) or method delegation to get the 'flow'. But this is async code, only matters *where* we run and *when*. The code we're looking at is an 'executor' for a (test) program.

Here I want `fn()` behavior which is sync and blocking, `defer(0, fn)` is not sync but cooperative (in the sense that other tasks in the queue can run before fn) but still blocking (in the sense that all these tasks marked as to-be-run-immediate run in a chain without gaps). `defer(10, fn)` is not sync and not blocking.

Please look at the `DeferrableTestSuite` in `suite.py`. We should remove all plain `yield`s in `run()`. What is your preferred way of writing this?

Comment (randy3k) [reply unittesting/core/st3/fast_runner.py]: It might come down to our different expectations about synchronization of the test cases.	
The major difference between the original runner and the new runner is whether we run the next test function/ test case in a deferred manner. It is also related to the plain `yield`s in `DeferrableTestSuite`.

```diff
@@ -48,7 +50,7 @@ class DeferringTextTestRunner(TextTestRunner):
                     startTestRun()
                 try:
                     deferred = test(result)
-                    defer(10, _continue_testing, deferred)
+                    _continue_testing(deferred)
```

When I first wrote it, I thought it would be just easier to defer everywhere. I have to admit that it may be not very efficient and strictly necessary. The idea here is to make sure the UI has time to respond and the tests won't block the UI. The name of the runner also suggests that it runs tests in a deferring manner. This is not guaranteed that the code will be synchronous.

I understand your point of keeping the code clean and simple. But if we look at the diff of the original runner and the fast runner, there are actually not many lines which are different. It is hard to imagine why we should not refactor the runners to make it more customizable. The real issue here is whether we should defer between tests. The setting `fast_timings` actually does not explain explicitly what it is being faster. Perhaps we need a more specific setting to control whether the coroutine is executed asynchronously and by how much. I am still a bit leaning towards `defer_timeout` with the understanding that `defer_timeout = 0` means `fn()` rather than `timeout(fn, 0)` in some of the lines.

The actual operation is actually hidden from the users. For the users, that number just means the number of seconds to wait. `0` just means "do not wait".

If you are worried about breakage, you have already protected yourself by writing the corresponding tests.

**Comment by randy3k** [comment]
TLDR, I am ok to replacing the old runner with your new runner. Even a little bit breakage is okay. TBH, not many people is using the deferred option. If necessary, I could do a quick search on GitHub and remind every repo with uses UnitTesting.

**Comment by randy3k** [comment]
How about making the new runner the default and rename the original runner LegacyDeferringTestRunner and add a setting `legacy_runner` with default false. I don’t really mind some breakage here, in fact I think most users will not be effected by this change

**Comment by kaste** [comment]
Okay, enough testing. There is something in master that makes 'circleci: build-linux' flaky. (Or the combination of this PR plus master.)

**Comment by randy3k** [comment]
it looks all good to me now. Merged now.

At some point, we might want to set the default value of `legacy_runner` to false.

**Comment by randy3k** [comment]
I will make a new release in a few days if everything runs smoothly

**Comment by kaste** [comment]
Yeah, that's the plan. Write a *good* update message. Essentially we remove *implicit* waiting time so users have to make their needed waiting time *explicit*. 'AWAIT_WORKER' _can_ be used here. Also users should almost always prefer `yield <condition>` over `yield <delay>`. Maybe deprecate until whatever April, 1st. Then flip the default, then April 15th remove the code. 

Is it we're lucky that 'circleci' is green on master, or did you change some setting here?

**Comment by randy3k** [comment]
I didn’t change any settings. Maybe you were being unlucky to catch some random glitches.
